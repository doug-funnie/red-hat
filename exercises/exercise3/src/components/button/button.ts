import glamorous from 'glamorous';
import { lighten } from 'polished';
import { ButtonHTMLAttributes } from 'react';
import { ThemedProps } from 'styles/theme';

type Props = Pick<ButtonHTMLAttributes<HTMLButtonElement>,
  'onClick'
  | 'disabled'
  | 'type'
>;

export const Button = glamorous.button<ThemedProps<Props>>(({ theme, disabled }) => {
  const bg = theme.colors.red;
  const activeBg = lighten(0.2, bg);

  return {
    border: 'none',
    borderRadius: 4,
    fontWeight: 700,
    height: 32,
    lineHeight: '32px',
    outline: 'none',
    padding: '0 10px',
    color: theme.colors.white,
    backgroundColor: bg,
    opacity: disabled ? 0.6 : 1,
    cursor: disabled ? 'not-allowed' : 'cursor',
    ':hover': {
      backgroundColor: activeBg,
    },
    ':active': {
      backgroundColor: activeBg,
    },
    ':focus': {
      backgroundColor: activeBg,
    },
  };
});

Button.defaultProps = {
  type: 'button',
};
