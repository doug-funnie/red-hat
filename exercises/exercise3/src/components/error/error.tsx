import { Div } from 'glamorous';
import * as React from 'react';
import { Button } from '../button';
import { H2, H3 } from '../text';

export interface Props {
  title?: string;
  message?: string;
  icon?: string;
  onRetry?(): void;
}

const defaultProps: Partial<Props> = {
  icon: 'ðŸ˜’',
  title: 'Oops!',
  message: 'Something went wrong.',
};

export const Error: React.SFC<Props> = ({
  title,
  message,
  icon,
  onRetry,
}) => (
  <Div display='flex' flexDirection='column' alignItems='center' justifyContent='center'>
    <Div fontSize={75}>{icon}</Div>
    <Div marginBottom={15} textAlign='center'>
      <H2>{title}</H2>
      <H3>{message}</H3>
    </Div>
    {!!onRetry && (
      <Button onClick={onRetry}>
        Give it another go
      </Button>
    )}
  </Div>
);

Error.defaultProps = defaultProps;
