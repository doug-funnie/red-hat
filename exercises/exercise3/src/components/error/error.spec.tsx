import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Button } from '../button';
import { Error, Props } from './error';

const props: Props = {
  title: 'Title',
  message: 'Message',
};

it('renders defaults', () => {
  const view = shallow(<Error />);
  expect(toJson(view)).toMatchSnapshot();
});

it('renders title and message props', () => {
  const view = shallow(<Error {...props} />);
  expect(toJson(view)).toMatchSnapshot();
});

it('calls on retry on button click', () => {
  const onRetryMock = jest.fn();
  const view = shallow(<Error {...props} onRetry={onRetryMock}/>);
  view.find(Button).simulate('click');
  expect(onRetryMock).toBeCalled();
});
