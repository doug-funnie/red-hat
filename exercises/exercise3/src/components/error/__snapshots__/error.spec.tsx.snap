// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders defaults 1`] = `
<glamorous.Div
  alignItems="center"
  display="flex"
  flexDirection="column"
  justifyContent="center"
>
  <glamorous.Div
    fontSize={75}
  >
    ðŸ˜’
  </glamorous.Div>
  <glamorous.Div
    marginBottom={15}
    textAlign="center"
  >
    <glamorous(h2)>
      Oops!
    </glamorous(h2)>
    <glamorous(h3)>
      Something went wrong.
    </glamorous(h3)>
  </glamorous.Div>
</glamorous.Div>
`;

exports[`renders title and message props 1`] = `
<glamorous.Div
  alignItems="center"
  display="flex"
  flexDirection="column"
  justifyContent="center"
>
  <glamorous.Div
    fontSize={75}
  >
    ðŸ˜’
  </glamorous.Div>
  <glamorous.Div
    marginBottom={15}
    textAlign="center"
  >
    <glamorous(h2)>
      Title
    </glamorous(h2)>
    <glamorous(h3)>
      Message
    </glamorous(h3)>
  </glamorous.Div>
</glamorous.Div>
`;
